--- a/ggml/src/ggml-cuda/conv-transpose-1d.cu
+++ b/ggml/src/ggml-cuda/conv-transpose-1d.cu
@@ -1,4 +1,5 @@
 #include "conv-transpose-1d.cuh"
+#include <algorithm>
 
 static  __global__ void conv_transpose_1d_kernel(
         const int s0, const int p0, const int d0, const int output_size,
@@ -6,11 +7,7 @@ static  __global__ void conv_transpose_1d_kernel(
         const int src1_ne0, const int src1_ne1, const int src1_ne2, const int src1_ne3,
         const int dst_ne0, const int dst_ne1, const int dst_ne2, const int dst_ne3,
         const float * src0, const float * src1,  float * dst) {
-    int global_index = threadIdx.x + blockIdx.x * blockDim.x;
-    if (global_index >= output_size) {
-        return;
-    }
-
+    for (int global_index = threadIdx.x + blockIdx.x * blockDim.x; global_index < output_size; global_index += blockDim.x * gridDim.x) {
     int out_index = global_index / dst_ne0;
 
     float accumulator = 0;
@@ -34,6 +31,7 @@ static  __global__ void conv_transpose_1d_kernel(
         }
     }
     dst[global_index] = accumulator;
+    }
     GGML_UNUSED_VARS(p0, d0, src0_ne3, src1_ne3, dst_ne3, src1_ne1, dst_ne1, src1_ne2, dst_ne2);
 }
 
@@ -45,7 +43,7 @@ static void conv_transpose_1d_f32_f32_cuda(
         const float * src0, const float * src1,  float * dst,
         cudaStream_t stream) {
 
-    const int num_blocks = (output_size + CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE - 1) / CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE;
+    const int num_blocks = std::min(65535, (output_size + CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE - 1) / CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE);
     conv_transpose_1d_kernel<<<num_blocks,CUDA_CONV_TRANPOSE_1D_BLOCK_SIZE, 0, stream>>>(
         s0,p0,d0,output_size,
         src0_ne0, src0_ne1,  src0_ne2, src0_ne3,
